Flight Academy Crashes with a shader compilation fail

### Title

Flight Academy - https://xemu.app/titles/41510002

### Bug Description

This game crashes with a shader compilation fail when trying to make it into menus

<details close>
<summary>Failing Fragment Shader from Flight Academy </summary>
<br>
#version 400

uniform vec2 clipRange;
uniform vec2 surfaceSize;

uniform vec4 c[192];

uniform vec4 fogColor;
uniform float fogParam[2];

#define fogPlane c[0x39]
#define texMat0 mat4(c[0x44], c[0x44+1], c[0x44+2], c[0x44+3])
#define texMat1 mat4(c[0x4c], c[0x4c+1], c[0x4c+2], c[0x4c+3])
#define texMat2 mat4(c[0x54], c[0x54+1], c[0x54+2], c[0x54+3])
#define texMat3 mat4(c[0x5c], c[0x5c+1], c[0x5c+2], c[0x5c+3])

vec4 oPos = vec4(0.0,0.0,0.0,1.0);
vec4 oD0 = vec4(0.0,0.0,0.0,1.0);
vec4 oD1 = vec4(0.0,0.0,0.0,1.0);
vec4 oB0 = vec4(0.0,0.0,0.0,1.0);
vec4 oB1 = vec4(0.0,0.0,0.0,1.0);
vec4 oPts = vec4(0.0,0.0,0.0,1.0);
vec4 oFog = vec4(1.0,0.0,0.0,1.0);
vec4 oT0 = vec4(0.0,0.0,0.0,1.0);
vec4 oT1 = vec4(0.0,0.0,0.0,1.0);
vec4 oT2 = vec4(0.0,0.0,0.0,1.0);
vec4 oT3 = vec4(0.0,0.0,0.0,1.0);

vec4 decompress_11_11_10(int cmp) {
    float x = float(bitfieldExtract(cmp, 0,  11)) / 1023.0;
    float y = float(bitfieldExtract(cmp, 11, 11)) / 1023.0;
    float z = float(bitfieldExtract(cmp, 22, 10)) / 511.0;
    return vec4(x, y, z, 1);
}
struct VertexData {
  float inv_w;
  vec4 D0;
  vec4 D1;
  vec4 B0;
  vec4 B1;
  float Fog;
  vec4 T0;
  vec4 T1;
  vec4 T2;
  vec4 T3;
};
noperspective out VertexData g_vtx;
#define vtx g_vtx

layout(location = 0) in vec4 v0;
layout(location = 1) in vec4 v1;
layout(location = 2) in vec4 v2;
layout(location = 3) in vec4 v3;
layout(location = 4) in vec4 v4;
layout(location = 5) in vec4 v5;
layout(location = 6) in vec4 v6;
layout(location = 7) in vec4 v7;
layout(location = 8) in vec4 v8;
layout(location = 9) in vec4 v9;
layout(location = 10) in vec4 v10;
layout(location = 11) in vec4 v11;
layout(location = 12) in vec4 v12;
layout(location = 13) in vec4 v13;
layout(location = 14) in vec4 v14;
layout(location = 15) in vec4 v15;

#define position      v0
#define weight        v1
#define normal        v2.xyz
#define diffuse       v3
#define specular      v4
#define fogCoord      v5.x
#define pointSize     v6
#define backDiffuse   v7
#define backSpecular  v8
#define texture0      v9
#define texture1      v10
#define texture2      v11
#define texture3      v12
#define reserved1     v13
#define reserved2     v14
#define reserved3     v15

uniform vec4 ltctxa[26];
uniform vec4 ltctxb[52];
uniform vec4 ltc1[20];

#define projectionMat mat4(c[0x04], c[0x04+1], c[0x04+2], c[0x04+3])
#define compositeMat mat4(c[0x00], c[0x00+1], c[0x00+2], c[0x00+3])

#define texPlaneS0 c[0x40 + 0]
#define texPlaneT0 c[0x40 + 1]
#define texPlaneQ0 c[0x40 + 2]
#define texPlaneR0 c[0x40 + 3]

#define texPlaneS1 c[0x48 + 0]
#define texPlaneT1 c[0x48 + 1]
#define texPlaneQ1 c[0x48 + 2]
#define texPlaneR1 c[0x48 + 3]

#define texPlaneS2 c[0x50 + 0]
#define texPlaneT2 c[0x50 + 1]
#define texPlaneQ2 c[0x50 + 2]
#define texPlaneR2 c[0x50 + 3]

#define texPlaneS3 c[0x58 + 0]
#define texPlaneT3 c[0x58 + 1]
#define texPlaneQ3 c[0x58 + 2]
#define texPlaneR3 c[0x58 + 3]

#define modelViewMat0 mat4(c[0x08], c[0x08+1], c[0x08+2], c[0x08+3])
#define modelViewMat1 mat4(c[0x10], c[0x10+1], c[0x10+2], c[0x10+3])
#define modelViewMat2 mat4(c[0x18], c[0x18+1], c[0x18+2], c[0x18+3])
#define modelViewMat3 mat4(c[0x20], c[0x20+1], c[0x20+2], c[0x20+3])

#define invModelViewMat0 mat4(c[0x0c], c[0x0c+1], c[0x0c+2], c[0x0c+3])
#define invModelViewMat1 mat4(c[0x14], c[0x14+1], c[0x14+2], c[0x14+3])
#define invModelViewMat2 mat4(c[0x1c], c[0x1c+1], c[0x1c+2], c[0x1c+3])
#define invModelViewMat3 mat4(c[0x24], c[0x24+1], c[0x24+2], c[0x24+3])

#define eyePosition c[0x38]

#define lightAmbientColor(i) ltctxb[0x00 + (i)*6].xyz
#define lightDiffuseColor(i) ltctxb[0x01 + (i)*6].xyz
#define lightSpecularColor(i) ltctxb[0x02 + (i)*6].xyz

#define lightSpotFalloff(i) ltctxa[0x00 + (i)*2].xyz
#define lightSpotDirection(i) ltctxa[0x01 + (i)*2]

#define lightLocalRange(i) ltc1[0x04 + (i)].x

#define sceneAmbientColor ltctxa[0x11].xyz
#define materialEmissionColor ltctxa[0x13].xyz

uniform mat4 invViewport;

void main() {
/* Skinning mode 0 */
vec4 tPosition = (position * modelViewMat0).xyzw;
vec3 tNormal = (vec4(normal, 0.0) * invModelViewMat0).xyz;
/* Texgen for stage 0 */
oT0.x = texture0.x;
oT0.y = texture0.y;
oT0.z = texture0.z;
oT0.w = texture0.w;
/* Texgen for stage 1 */
oT1.x = texture1.x;
oT1.y = texture1.y;
oT1.z = texture1.z;
oT1.w = texture1.w;
/* Texgen for stage 2 */
oT2.x = texture2.x;
oT2.y = texture2.y;
oT2.z = texture2.z;
oT2.w = texture2.w;
/* Texgen for stage 3 */
oT3.x = texture3.x;
oT3.y = texture3.y;
oT3.z = texture3.z;
oT3.w = texture3.w;
oT0 = oT0 * texMat0;
  oD0 = diffuse;
  oD1 = specular;
  oB0 = backDiffuse;
  oB1 = backSpecular;
  tPosition = position;
   oPos = invViewport * (tPosition * compositeMat);
   oPos.z = oPos.z * 2.0 - oPos.w;
  float d_e = length(position * modelViewMat0);
  oPts.x = 1/sqrt(nan + nan*d_e + inf*d_e*d_e) + -0.032258;
  oPts.x = min(oPts.x*62.000000 + 2.000000, 64.0) * 1;
  vtx.inv_w = 1.0 / oPos.w;
  oFog.xyzw = vec4(1.0);

  vtx.D0 = clamp(oD0, 0.0, 1.0) * vtx.inv_w;
  vtx.D1 = clamp(oD1, 0.0, 1.0) * vtx.inv_w;
  vtx.B0 = clamp(oB0, 0.0, 1.0) * vtx.inv_w;
  vtx.B1 = clamp(oB1, 0.0, 1.0) * vtx.inv_w;
  vtx.Fog = oFog.x * vtx.inv_w;
  vtx.T0 = oT0 * vtx.inv_w;
  vtx.T1 = oT1 * vtx.inv_w;
  vtx.T2 = oT2 * vtx.inv_w;
  vtx.T3 = oT3 * vtx.inv_w;
  gl_Position = oPos;
  gl_PointSize = oPts.x;

}


nv2a: vertex shader compilation failed: 0(170) : error C1503: undefined variable "nan"
0(170) : error C1503: undefined variable "nan"
0(170) : error C1503: undefined variable "inf"
</details>

### Expected Behavior

shouldn't crash



### xemu Version

Version: 0.7.60
Branch: master
Commit: https://github.com/mborgerson/xemu/commit/8d274e5f9a9dbe777ffba6a8b3d269adbf682690
Date: Wed Jul 13 23:19:36 UTC 2022

### System Information

<html><body>
<!--StartFragment-->

Field | Value
-- | --
OS | Windows 10
CPU | AMD Ryzen 5 2600 Six-Core Processor
Graphics Device | NVIDIA GeForce RTX 3060 Ti/PCIe/SSE2
Graphics Driver | 4.0.0 NVIDIA 512.95

<!--EndFragment-->
</body>
</html>

### Additional Context

_No response_

--- a/hw/xbox/nv2a/pgraph/gl/renderer.h
+++ b/hw/xbox/nv2a/pgraph/gl/renderer.h
@@ -126,6 +126,7 @@ typedef struct ShaderBinding {
 
     GLint clip_region_loc[8];
 
+    GLint point_params_loc[8];
     GLint material_alpha_loc;
 } ShaderBinding;
 

--- a/hw/xbox/nv2a/pgraph/gl/shaders.c
+++ b/hw/xbox/nv2a/pgraph/gl/shaders.c
@@ -190,6 +190,11 @@ static void update_shader_constant_locations(ShaderBinding *binding)
         binding->clip_region_loc[i] = glGetUniformLocation(binding->gl_program, tmp);
     }
 
+    for (int i = 0; i < 8; ++i) {
+        snprintf(tmp, sizeof(tmp), "pointParams[%d]", i);
+        binding->point_params_loc[i] = glGetUniformLocation(binding->gl_program, tmp);
+    }
+
     if (binding->state.fixed_function) {
         binding->material_alpha_loc =
             glGetUniformLocation(binding->gl_program, "material_alpha");
@@ -948,6 +953,13 @@ static void shader_update_constants(PGRAPHState *pg, ShaderBinding *binding,
 
         glUniform4i(r->shader_binding->clip_region_loc[i],
                     x_min, y_min_xlat, x_max, y_max_xlat);
+    }
+
+    for (i = 0; i < 8; ++i) {
+        GLint loc = binding->point_params_loc[i];
+        if (loc != -1) {
+            glUniform1f(loc, pg->point_params[i]);
+        }
     }
 
     if (binding->material_alpha_loc != -1) {

--- a/hw/xbox/nv2a/pgraph/glsl/vsh-ff.c
+++ b/hw/xbox/nv2a/pgraph/glsl/vsh-ff.c
@@ -482,14 +482,12 @@ GLSL_DEFINE(materialEmissionColor, GLSL_LTCTXA(NV_IGRAPH_XF_LTCTXA_CM_COL) ".xyz
 
     /* FIXME: Testing */
     if (state->point_params_enable) {
-        mstring_append_fmt(
+        mstring_append_fmt(uniforms, "%sfloat pointParams[8];\n", u);
+        mstring_append(
             body,
             "  float d_e = length(position * modelViewMat0);\n"
-            "  oPts.x = 1/sqrt(%f + %f*d_e + %f*d_e*d_e) + %f;\n",
-            state->point_params[0], state->point_params[1], state->point_params[2],
-            state->point_params[6]);
-        mstring_append_fmt(body, "  oPts.x = min(oPts.x*%f + %f, 64.0) * %d;\n",
-                           state->point_params[3], state->point_params[7],
+            "  oPts.x = 1/sqrt(pointParams[0] + pointParams[1] * d_e + pointParams[2] * d_e * d_e) + pointParams[6];\n");
+        mstring_append_fmt(body, "  oPts.x = min(oPts.x * pointParams[3] + pointParams[7], 64.0) * %d;\n",
                            state->surface_scale_factor);
     } else {
         mstring_append_fmt(body, "  oPts.x = %f * %d;\n", state->point_size,

--- a/hw/xbox/nv2a/pgraph/vk/renderer.h
+++ b/hw/xbox/nv2a/pgraph/vk/renderer.h
@@ -194,6 +194,7 @@ typedef struct ShaderBinding {
 
     int clip_region_loc;
 
+    int point_params_loc[8];
     int material_alpha_loc;
 
     int uniform_attrs_loc;

--- a/hw/xbox/nv2a/pgraph/vk/renderer.h
+++ b/hw/xbox/nv2a/pgraph/vk/renderer.h
@@ -191,10 +191,9 @@ typedef struct ShaderBinding {
     int light_local_position_loc[NV2A_MAX_LIGHTS];
     int light_local_attenuation_loc[NV2A_MAX_LIGHTS];
     int specular_power_loc;
+    int point_params_loc;
 
     int clip_region_loc;
-
-    int point_params_loc[8];
     int material_alpha_loc;
 
     int uniform_attrs_loc;

--- a/hw/xbox/nv2a/pgraph/vk/shaders.c
+++ b/hw/xbox/nv2a/pgraph/vk/shaders.c
@@ -308,11 +308,8 @@ static void update_shader_constant_locations(ShaderBinding *binding)
     binding->clip_region_loc =
         uniform_index(&binding->fragment->uniforms, "clipRegion");
 
-    for (int i = 0; i < 8; ++i) {
-        snprintf(tmp, sizeof(tmp), "pointParams[%d]", i);
-        binding->point_params_loc[i] =
-            uniform_index(&binding->vertex->uniforms, tmp);
-    }
+    binding->point_params_loc =
+        uniform_index(&binding->vertex->uniforms, "pointParams");
 
     binding->material_alpha_loc =
         uniform_index(&binding->vertex->uniforms, "material_alpha");
@@ -726,11 +723,9 @@ static void shader_update_constants(PGRAPHState *pg, ShaderBinding *binding,
     uniform1iv(&binding->fragment->uniforms, binding->clip_region_loc,
                      8 * 4, (void *)clip_regions);
 
-    for (int i = 0; i < 8; ++i) {
-        int loc = binding->point_params_loc[i];
-        if (loc != -1) {
-            uniform1f(&binding->vertex->uniforms, loc, pg->point_params[i]);
-        }
+    if (binding->point_params_loc != -1) {
+        uniform1iv(&binding->vertex->uniforms, binding->point_params_loc,
+                   ARRAY_SIZE(pg->point_params), (void *)pg->point_params);
     }
 
     if (binding->material_alpha_loc != -1) {

--- a/hw/xbox/nv2a/pgraph/vk/shaders.c
+++ b/hw/xbox/nv2a/pgraph/vk/shaders.c
@@ -308,6 +308,12 @@ static void update_shader_constant_locations(ShaderBinding *binding)
     binding->clip_region_loc =
         uniform_index(&binding->fragment->uniforms, "clipRegion");
 
+    for (int i = 0; i < 8; ++i) {
+        snprintf(tmp, sizeof(tmp), "pointParams[%d]", i);
+        binding->point_params_loc[i] =
+            uniform_index(&binding->vertex->uniforms, tmp);
+    }
+
     binding->material_alpha_loc =
         uniform_index(&binding->vertex->uniforms, "material_alpha");
 
@@ -719,6 +725,13 @@ static void shader_update_constants(PGRAPHState *pg, ShaderBinding *binding,
     }
     uniform1iv(&binding->fragment->uniforms, binding->clip_region_loc,
                      8 * 4, (void *)clip_regions);
+
+    for (int i = 0; i < 8; ++i) {
+        int loc = binding->point_params_loc[i];
+        if (loc != -1) {
+            uniform1f(&binding->vertex->uniforms, loc, pg->point_params[i]);
+        }
+    }
 
     if (binding->material_alpha_loc != -1) {
         uniform1f(&binding->vertex->uniforms, binding->material_alpha_loc,

